{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f69c9b0c",
   "metadata": {},
   "source": [
    "# Python Review Project\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "id": "87298971",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current working directory: /mnt/c/Users/Anton/Documents/QBIO490/qbio_490_antonwagner/analysis_data\n",
      "New working directory: /mnt/c/Users/Anton/Documents/QBIO490/qbio_490_antonwagner/analysis_data\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "print(\"Current working directory: {0}\".format(os.getcwd()))\n",
    "\n",
    "os.chdir('../analysis_data')\n",
    "\n",
    "print(\"New working directory: {0}\".format(os.getcwd()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "787db162",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Below are the dataframes contained in this dataset and their dimensions:\n",
      "\n",
      "clinical\n",
      "\t194 rows\n",
      "\t171 columns\n",
      "CNV\n",
      "\t110 rows\n",
      "\t19285 columns\n",
      "followup\n",
      "\t352 rows\n",
      "\t27 columns\n",
      "medical_history\n",
      "\t370 rows\n",
      "\t4 columns\n",
      "methylation\n",
      "\t107 rows\n",
      "\t15885 columns\n",
      "phosphoproteomics\n",
      "\t194 rows\n",
      "\t81550 columns\n",
      "phosphoproteomics_gene\n",
      "\t194 rows\n",
      "\t6127 columns\n",
      "proteomics\n",
      "\t194 rows\n",
      "\t11710 columns\n",
      "somatic_mutation\n",
      "\t8350 rows\n",
      "\t3 columns\n",
      "transcriptomics\n",
      "\t185 rows\n",
      "\t19275 columns\n"
     ]
    }
   ],
   "source": [
    "# Set up CPTAC\n",
    "\n",
    "#import cptac\n",
    "import cptac\n",
    "\n",
    "# View datadets\n",
    "cptac.list_datasets()\n",
    "\n",
    "# Download the clear cell renal cell carcinoma (kidney) data set\n",
    "cptac.download(dataset=\"Ccrcc\")\n",
    "ccrcc = cptac.Ccrcc()\n",
    "\n",
    "ccrcc.list_data()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "id": "6e24aa2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np # support for arrays and vectorized operations\n",
    "import pandas as pd # support for DataFrames and Series\n",
    "import matplotlib.pyplot as plt # main graphing library\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.neighbors import KNeighborsClassifier # default number of neighbors looked at is 5\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "id": "71a3cda2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(194, 11710)\n"
     ]
    }
   ],
   "source": [
    "# get protein data\n",
    "\n",
    "protein_data = ccrcc.get_proteomics() # get the proteomics data\n",
    "\n",
    "protein_data.columns = protein_data.columns.get_level_values(0) \n",
    "\n",
    "protein_data # view the data\n",
    "print(protein_data.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "6e2c235d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(185, 19275)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00004</th>\n",
       "      <td>0.995336</td>\n",
       "      <td>16.677828</td>\n",
       "      <td>353.263362</td>\n",
       "      <td>0.046634</td>\n",
       "      <td>0.031027</td>\n",
       "      <td>17.196885</td>\n",
       "      <td>0.896109</td>\n",
       "      <td>15.831130</td>\n",
       "      <td>2.938550</td>\n",
       "      <td>0.019029</td>\n",
       "      <td>...</td>\n",
       "      <td>2.622184</td>\n",
       "      <td>5.415676</td>\n",
       "      <td>5.293887</td>\n",
       "      <td>3.731920</td>\n",
       "      <td>12.541358</td>\n",
       "      <td>2.677696</td>\n",
       "      <td>15.595194</td>\n",
       "      <td>11.401808</td>\n",
       "      <td>9.976986</td>\n",
       "      <td>23.334614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00010</th>\n",
       "      <td>0.679400</td>\n",
       "      <td>16.682712</td>\n",
       "      <td>359.078446</td>\n",
       "      <td>0.077350</td>\n",
       "      <td>0.068617</td>\n",
       "      <td>13.560508</td>\n",
       "      <td>1.743989</td>\n",
       "      <td>16.690257</td>\n",
       "      <td>3.154143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.873604</td>\n",
       "      <td>9.209695</td>\n",
       "      <td>3.669353</td>\n",
       "      <td>2.560578</td>\n",
       "      <td>13.570779</td>\n",
       "      <td>4.097483</td>\n",
       "      <td>15.449647</td>\n",
       "      <td>11.550727</td>\n",
       "      <td>9.432121</td>\n",
       "      <td>25.724814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00011</th>\n",
       "      <td>0.354549</td>\n",
       "      <td>0.245606</td>\n",
       "      <td>222.075350</td>\n",
       "      <td>0.060736</td>\n",
       "      <td>0.273536</td>\n",
       "      <td>1.321499</td>\n",
       "      <td>0.172369</td>\n",
       "      <td>18.757568</td>\n",
       "      <td>6.942752</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>7.998655</td>\n",
       "      <td>28.780560</td>\n",
       "      <td>2.801800</td>\n",
       "      <td>2.503315</td>\n",
       "      <td>10.209840</td>\n",
       "      <td>0.178842</td>\n",
       "      <td>11.670596</td>\n",
       "      <td>11.342045</td>\n",
       "      <td>6.763858</td>\n",
       "      <td>32.090615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00026</th>\n",
       "      <td>2.543775</td>\n",
       "      <td>16.347532</td>\n",
       "      <td>228.282343</td>\n",
       "      <td>0.085684</td>\n",
       "      <td>0.152020</td>\n",
       "      <td>7.868391</td>\n",
       "      <td>1.448911</td>\n",
       "      <td>17.648610</td>\n",
       "      <td>6.175010</td>\n",
       "      <td>0.031078</td>\n",
       "      <td>...</td>\n",
       "      <td>2.754936</td>\n",
       "      <td>12.639323</td>\n",
       "      <td>5.262024</td>\n",
       "      <td>2.796869</td>\n",
       "      <td>10.718552</td>\n",
       "      <td>0.800663</td>\n",
       "      <td>15.887414</td>\n",
       "      <td>11.788588</td>\n",
       "      <td>8.169953</td>\n",
       "      <td>24.752283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00079</th>\n",
       "      <td>4.355205</td>\n",
       "      <td>4.858958</td>\n",
       "      <td>275.090167</td>\n",
       "      <td>0.106359</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.863003</td>\n",
       "      <td>2.338081</td>\n",
       "      <td>15.480282</td>\n",
       "      <td>4.584445</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>7.497914</td>\n",
       "      <td>14.400917</td>\n",
       "      <td>2.907591</td>\n",
       "      <td>2.417113</td>\n",
       "      <td>10.127549</td>\n",
       "      <td>3.442177</td>\n",
       "      <td>12.807428</td>\n",
       "      <td>17.494840</td>\n",
       "      <td>9.733803</td>\n",
       "      <td>24.528238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01646.N</th>\n",
       "      <td>1.287655</td>\n",
       "      <td>11.076241</td>\n",
       "      <td>228.037113</td>\n",
       "      <td>0.106505</td>\n",
       "      <td>0.066692</td>\n",
       "      <td>8.195971</td>\n",
       "      <td>0.982361</td>\n",
       "      <td>18.086400</td>\n",
       "      <td>10.225658</td>\n",
       "      <td>0.040903</td>\n",
       "      <td>...</td>\n",
       "      <td>2.338694</td>\n",
       "      <td>3.274051</td>\n",
       "      <td>3.500041</td>\n",
       "      <td>3.223773</td>\n",
       "      <td>15.663483</td>\n",
       "      <td>3.219438</td>\n",
       "      <td>17.530849</td>\n",
       "      <td>6.471017</td>\n",
       "      <td>11.860328</td>\n",
       "      <td>25.489666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01648.N</th>\n",
       "      <td>1.435986</td>\n",
       "      <td>9.772280</td>\n",
       "      <td>291.393930</td>\n",
       "      <td>0.172457</td>\n",
       "      <td>0.214180</td>\n",
       "      <td>8.955806</td>\n",
       "      <td>1.838108</td>\n",
       "      <td>14.597847</td>\n",
       "      <td>6.598100</td>\n",
       "      <td>0.056297</td>\n",
       "      <td>...</td>\n",
       "      <td>1.983803</td>\n",
       "      <td>1.486905</td>\n",
       "      <td>3.865995</td>\n",
       "      <td>3.197000</td>\n",
       "      <td>15.581449</td>\n",
       "      <td>5.774722</td>\n",
       "      <td>17.131855</td>\n",
       "      <td>10.133624</td>\n",
       "      <td>15.256983</td>\n",
       "      <td>22.669748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01649.N</th>\n",
       "      <td>1.082318</td>\n",
       "      <td>8.378616</td>\n",
       "      <td>249.779349</td>\n",
       "      <td>0.100673</td>\n",
       "      <td>0.160751</td>\n",
       "      <td>7.419323</td>\n",
       "      <td>2.290458</td>\n",
       "      <td>17.666594</td>\n",
       "      <td>11.108717</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.007450</td>\n",
       "      <td>2.590684</td>\n",
       "      <td>3.805016</td>\n",
       "      <td>2.943559</td>\n",
       "      <td>17.274857</td>\n",
       "      <td>5.856484</td>\n",
       "      <td>16.392044</td>\n",
       "      <td>6.775269</td>\n",
       "      <td>12.881048</td>\n",
       "      <td>25.672959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01651.N</th>\n",
       "      <td>0.770924</td>\n",
       "      <td>15.539566</td>\n",
       "      <td>273.743429</td>\n",
       "      <td>0.208978</td>\n",
       "      <td>0.185383</td>\n",
       "      <td>12.206999</td>\n",
       "      <td>2.377283</td>\n",
       "      <td>13.946563</td>\n",
       "      <td>6.920178</td>\n",
       "      <td>0.022739</td>\n",
       "      <td>...</td>\n",
       "      <td>1.861395</td>\n",
       "      <td>2.904921</td>\n",
       "      <td>3.089317</td>\n",
       "      <td>2.529058</td>\n",
       "      <td>14.839389</td>\n",
       "      <td>5.801733</td>\n",
       "      <td>18.372506</td>\n",
       "      <td>9.149521</td>\n",
       "      <td>13.918938</td>\n",
       "      <td>24.578885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01808.N</th>\n",
       "      <td>0.622108</td>\n",
       "      <td>16.125722</td>\n",
       "      <td>172.780700</td>\n",
       "      <td>0.091174</td>\n",
       "      <td>0.025541</td>\n",
       "      <td>9.702769</td>\n",
       "      <td>2.006484</td>\n",
       "      <td>17.649919</td>\n",
       "      <td>10.172484</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.713981</td>\n",
       "      <td>3.014346</td>\n",
       "      <td>3.411386</td>\n",
       "      <td>3.001192</td>\n",
       "      <td>18.717445</td>\n",
       "      <td>12.680195</td>\n",
       "      <td>19.327095</td>\n",
       "      <td>6.351638</td>\n",
       "      <td>14.957318</td>\n",
       "      <td>24.678258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>185 rows × 19275 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name             A1BG       A1CF         A2M     A2ML1   A3GALT2     A4GALT  \\\n",
       "Patient_ID                                                                    \n",
       "C3L-00004    0.995336  16.677828  353.263362  0.046634  0.031027  17.196885   \n",
       "C3L-00010    0.679400  16.682712  359.078446  0.077350  0.068617  13.560508   \n",
       "C3L-00011    0.354549   0.245606  222.075350  0.060736  0.273536   1.321499   \n",
       "C3L-00026    2.543775  16.347532  228.282343  0.085684  0.152020   7.868391   \n",
       "C3L-00079    4.355205   4.858958  275.090167  0.106359  0.000000   6.863003   \n",
       "...               ...        ...         ...       ...       ...        ...   \n",
       "C3N-01646.N  1.287655  11.076241  228.037113  0.106505  0.066692   8.195971   \n",
       "C3N-01648.N  1.435986   9.772280  291.393930  0.172457  0.214180   8.955806   \n",
       "C3N-01649.N  1.082318   8.378616  249.779349  0.100673  0.160751   7.419323   \n",
       "C3N-01651.N  0.770924  15.539566  273.743429  0.208978  0.185383  12.206999   \n",
       "C3N-01808.N  0.622108  16.125722  172.780700  0.091174  0.025541   9.702769   \n",
       "\n",
       "Name            A4GNT       AAAS       AACS     AADAC  ...    ZWILCH  \\\n",
       "Patient_ID                                             ...             \n",
       "C3L-00004    0.896109  15.831130   2.938550  0.019029  ...  2.622184   \n",
       "C3L-00010    1.743989  16.690257   3.154143  0.000000  ...  2.873604   \n",
       "C3L-00011    0.172369  18.757568   6.942752  0.000000  ...  7.998655   \n",
       "C3L-00026    1.448911  17.648610   6.175010  0.031078  ...  2.754936   \n",
       "C3L-00079    2.338081  15.480282   4.584445  0.000000  ...  7.497914   \n",
       "...               ...        ...        ...       ...  ...       ...   \n",
       "C3N-01646.N  0.982361  18.086400  10.225658  0.040903  ...  2.338694   \n",
       "C3N-01648.N  1.838108  14.597847   6.598100  0.056297  ...  1.983803   \n",
       "C3N-01649.N  2.290458  17.666594  11.108717  0.000000  ...  2.007450   \n",
       "C3N-01651.N  2.377283  13.946563   6.920178  0.022739  ...  1.861395   \n",
       "C3N-01808.N  2.006484  17.649919  10.172484  0.000000  ...  1.713981   \n",
       "\n",
       "Name             ZWINT      ZXDA      ZXDB       ZXDC     ZYG11A     ZYG11B  \\\n",
       "Patient_ID                                                                    \n",
       "C3L-00004     5.415676  5.293887  3.731920  12.541358   2.677696  15.595194   \n",
       "C3L-00010     9.209695  3.669353  2.560578  13.570779   4.097483  15.449647   \n",
       "C3L-00011    28.780560  2.801800  2.503315  10.209840   0.178842  11.670596   \n",
       "C3L-00026    12.639323  5.262024  2.796869  10.718552   0.800663  15.887414   \n",
       "C3L-00079    14.400917  2.907591  2.417113  10.127549   3.442177  12.807428   \n",
       "...                ...       ...       ...        ...        ...        ...   \n",
       "C3N-01646.N   3.274051  3.500041  3.223773  15.663483   3.219438  17.530849   \n",
       "C3N-01648.N   1.486905  3.865995  3.197000  15.581449   5.774722  17.131855   \n",
       "C3N-01649.N   2.590684  3.805016  2.943559  17.274857   5.856484  16.392044   \n",
       "C3N-01651.N   2.904921  3.089317  2.529058  14.839389   5.801733  18.372506   \n",
       "C3N-01808.N   3.014346  3.411386  3.001192  18.717445  12.680195  19.327095   \n",
       "\n",
       "Name               ZYX      ZZEF1       ZZZ3  \n",
       "Patient_ID                                    \n",
       "C3L-00004    11.401808   9.976986  23.334614  \n",
       "C3L-00010    11.550727   9.432121  25.724814  \n",
       "C3L-00011    11.342045   6.763858  32.090615  \n",
       "C3L-00026    11.788588   8.169953  24.752283  \n",
       "C3L-00079    17.494840   9.733803  24.528238  \n",
       "...                ...        ...        ...  \n",
       "C3N-01646.N   6.471017  11.860328  25.489666  \n",
       "C3N-01648.N  10.133624  15.256983  22.669748  \n",
       "C3N-01649.N   6.775269  12.881048  25.672959  \n",
       "C3N-01651.N   9.149521  13.918938  24.578885  \n",
       "C3N-01808.N   6.351638  14.957318  24.678258  \n",
       "\n",
       "[185 rows x 19275 columns]"
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get RNA data\n",
    "\n",
    "rna_data = ccrcc.get_transcriptomics()\n",
    "print(rna_data.shape)\n",
    "rna_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "id": "f5a2910c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(194, 171)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>Sample_Tumor_Normal</th>\n",
       "      <th>tumor/normal</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>height_in_cm</th>\n",
       "      <th>height_in_inch</th>\n",
       "      <th>weight_in_kg</th>\n",
       "      <th>weight_in_lb</th>\n",
       "      <th>BMI</th>\n",
       "      <th>race</th>\n",
       "      <th>...</th>\n",
       "      <th>histologic_type_of_normal_tissue</th>\n",
       "      <th>slide_is_free_of_tumor</th>\n",
       "      <th>consistent_with_local_pathology_report</th>\n",
       "      <th>findings_not_consistent_with_local_pathology_report</th>\n",
       "      <th>weight_in_mg</th>\n",
       "      <th>minutes_clamp_1_to_collection</th>\n",
       "      <th>minutes_clamp_2_to_collection</th>\n",
       "      <th>minutes_collection_to_frozen</th>\n",
       "      <th>consistent_with_diagnostic_report</th>\n",
       "      <th>patient_medications</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00004</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>TN</td>\n",
       "      <td>Male</td>\n",
       "      <td>72</td>\n",
       "      <td>170.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>22.80</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>484.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21</td>\n",
       "      <td>Yes</td>\n",
       "      <td>rivaroxaban,tylenol,Aspirin,glycolax,colace,zocor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00010</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>TN</td>\n",
       "      <td>Male</td>\n",
       "      <td>30</td>\n",
       "      <td>177.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>107.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>34.15</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>575.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Rivaroxaban,Esomeprazole ,Tramadol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00011</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>TN</td>\n",
       "      <td>Female</td>\n",
       "      <td>63</td>\n",
       "      <td>180.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>196.0</td>\n",
       "      <td>27.47</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>272.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Multi Vitamin,Levothyroxine Sodium,Ibandronate...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00026</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>TN</td>\n",
       "      <td>Female</td>\n",
       "      <td>65</td>\n",
       "      <td>163.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>165.0</td>\n",
       "      <td>28.23</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>212.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Levothyroxine Sodium,Cyproheptadine HCL,Citrac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00079</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>TN</td>\n",
       "      <td>Male</td>\n",
       "      <td>49</td>\n",
       "      <td>175.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>256.0</td>\n",
       "      <td>37.88</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>675.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28</td>\n",
       "      <td>Yes</td>\n",
       "      <td>ibuprofen,Norco,miralax</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01646.N</th>\n",
       "      <td>Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>renal cortex, medulla</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>256.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01648.N</th>\n",
       "      <td>Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>renal cortex</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>288.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01649.N</th>\n",
       "      <td>Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>renal medulla, pelvis</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>301.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01651.N</th>\n",
       "      <td>Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>renal cortex, medulla</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>280.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3N-01808.N</th>\n",
       "      <td>Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>renal cortex</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>195.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>194 rows × 171 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name        Sample_Tumor_Normal tumor/normal  gender  age  height_in_cm  \\\n",
       "Patient_ID                                                                \n",
       "C3L-00004                 Tumor           TN    Male   72         170.0   \n",
       "C3L-00010                 Tumor           TN    Male   30         177.0   \n",
       "C3L-00011                 Tumor           TN  Female   63         180.0   \n",
       "C3L-00026                 Tumor           TN  Female   65         163.0   \n",
       "C3L-00079                 Tumor           TN    Male   49         175.0   \n",
       "...                         ...          ...     ...  ...           ...   \n",
       "C3N-01646.N              Normal          NaN     NaN  NaN           NaN   \n",
       "C3N-01648.N              Normal          NaN     NaN  NaN           NaN   \n",
       "C3N-01649.N              Normal          NaN     NaN  NaN           NaN   \n",
       "C3N-01651.N              Normal          NaN     NaN  NaN           NaN   \n",
       "C3N-01808.N              Normal          NaN     NaN  NaN           NaN   \n",
       "\n",
       "Name         height_in_inch  weight_in_kg  weight_in_lb    BMI   race  ...  \\\n",
       "Patient_ID                                                             ...   \n",
       "C3L-00004              67.0          66.0         145.0  22.80  White  ...   \n",
       "C3L-00010              70.0         107.0         236.0  34.15  White  ...   \n",
       "C3L-00011              71.0          89.0         196.0  27.47  White  ...   \n",
       "C3L-00026              64.0          75.0         165.0  28.23  White  ...   \n",
       "C3L-00079              69.0         116.0         256.0  37.88  White  ...   \n",
       "...                     ...           ...           ...    ...    ...  ...   \n",
       "C3N-01646.N             NaN           NaN           NaN    NaN    NaN  ...   \n",
       "C3N-01648.N             NaN           NaN           NaN    NaN    NaN  ...   \n",
       "C3N-01649.N             NaN           NaN           NaN    NaN    NaN  ...   \n",
       "C3N-01651.N             NaN           NaN           NaN    NaN    NaN  ...   \n",
       "C3N-01808.N             NaN           NaN           NaN    NaN    NaN  ...   \n",
       "\n",
       "Name        histologic_type_of_normal_tissue slide_is_free_of_tumor  \\\n",
       "Patient_ID                                                            \n",
       "C3L-00004                                NaN                    NaN   \n",
       "C3L-00010                                NaN                    NaN   \n",
       "C3L-00011                                NaN                    NaN   \n",
       "C3L-00026                                NaN                    NaN   \n",
       "C3L-00079                                NaN                    NaN   \n",
       "...                                      ...                    ...   \n",
       "C3N-01646.N            renal cortex, medulla                    Yes   \n",
       "C3N-01648.N                     renal cortex                    Yes   \n",
       "C3N-01649.N            renal medulla, pelvis                    Yes   \n",
       "C3N-01651.N            renal cortex, medulla                    Yes   \n",
       "C3N-01808.N                     renal cortex                    Yes   \n",
       "\n",
       "Name        consistent_with_local_pathology_report  \\\n",
       "Patient_ID                                           \n",
       "C3L-00004                                      Yes   \n",
       "C3L-00010                                      Yes   \n",
       "C3L-00011                                      Yes   \n",
       "C3L-00026                                      Yes   \n",
       "C3L-00079                                      Yes   \n",
       "...                                            ...   \n",
       "C3N-01646.N                                    Yes   \n",
       "C3N-01648.N                                    Yes   \n",
       "C3N-01649.N                                    Yes   \n",
       "C3N-01651.N                                    Yes   \n",
       "C3N-01808.N                                    Yes   \n",
       "\n",
       "Name        findings_not_consistent_with_local_pathology_report  weight_in_mg  \\\n",
       "Patient_ID                                                                      \n",
       "C3L-00004                                                  NaN          484.0   \n",
       "C3L-00010                                                  NaN          575.0   \n",
       "C3L-00011                                                  NaN          272.0   \n",
       "C3L-00026                                                  NaN          212.0   \n",
       "C3L-00079                                                  NaN          675.0   \n",
       "...                                                        ...            ...   \n",
       "C3N-01646.N                                                NaN          256.0   \n",
       "C3N-01648.N                                                NaN          288.0   \n",
       "C3N-01649.N                                                NaN          301.0   \n",
       "C3N-01651.N                                                NaN          280.0   \n",
       "C3N-01808.N                                                NaN          195.0   \n",
       "\n",
       "Name        minutes_clamp_1_to_collection minutes_clamp_2_to_collection  \\\n",
       "Patient_ID                                                                \n",
       "C3L-00004                             NaN                           NaN   \n",
       "C3L-00010                             NaN                           NaN   \n",
       "C3L-00011                             NaN                           NaN   \n",
       "C3L-00026                            13.0                           NaN   \n",
       "C3L-00079                             NaN                           NaN   \n",
       "...                                   ...                           ...   \n",
       "C3N-01646.N                           NaN                           NaN   \n",
       "C3N-01648.N                           NaN                           NaN   \n",
       "C3N-01649.N                           NaN                           NaN   \n",
       "C3N-01651.N                           NaN                           NaN   \n",
       "C3N-01808.N                           NaN                           NaN   \n",
       "\n",
       "Name        minutes_collection_to_frozen consistent_with_diagnostic_report  \\\n",
       "Patient_ID                                                                   \n",
       "C3L-00004                             21                               Yes   \n",
       "C3L-00010                             17                               Yes   \n",
       "C3L-00011                             18                               Yes   \n",
       "C3L-00026                             13                               Yes   \n",
       "C3L-00079                             28                               Yes   \n",
       "...                                  ...                               ...   \n",
       "C3N-01646.N                           20                               NaN   \n",
       "C3N-01648.N                            8                               NaN   \n",
       "C3N-01649.N                           29                               NaN   \n",
       "C3N-01651.N                           22                               NaN   \n",
       "C3N-01808.N                           29                               NaN   \n",
       "\n",
       "Name                                       patient_medications  \n",
       "Patient_ID                                                      \n",
       "C3L-00004    rivaroxaban,tylenol,Aspirin,glycolax,colace,zocor  \n",
       "C3L-00010                   Rivaroxaban,Esomeprazole ,Tramadol  \n",
       "C3L-00011    Multi Vitamin,Levothyroxine Sodium,Ibandronate...  \n",
       "C3L-00026    Levothyroxine Sodium,Cyproheptadine HCL,Citrac...  \n",
       "C3L-00079                              ibuprofen,Norco,miralax  \n",
       "...                                                        ...  \n",
       "C3N-01646.N                                                NaN  \n",
       "C3N-01648.N                                                NaN  \n",
       "C3N-01649.N                                                NaN  \n",
       "C3N-01651.N                                                NaN  \n",
       "C3N-01808.N                                                NaN  \n",
       "\n",
       "[194 rows x 171 columns]"
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get clinical data\n",
    "\n",
    "clinical_data = ccrcc.get_clinical()\n",
    "print(clinical_data.shape)\n",
    "clinical_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "8ae0f55d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top 5 DE RNAs:  ['AJAP1', 'DPEP1', 'GALNT5', 'IL20RB', 'RYR2']\n",
      "DE values:  [1.7385942919844544, 1.5409691011644857, 1.5137845013254805, 1.4021781083904463, 1.395319936401435]\n"
     ]
    }
   ],
   "source": [
    "# Find the top 5 DE RNA\n",
    "\n",
    "# mask out patient IDs who do not have rna_data\n",
    "clinical_data_with_rna = clinical_data.loc[rna_data.index, :]\n",
    "\n",
    "# create masks for the two clinical feature groups\n",
    "stageI_mask_with_rna = np.where(clinical_data_with_rna.loc[:, \"tumor_stage_pathological\"] == \"Stage I\", True, False)\n",
    "stageIII_mask_with_rna = np.where(clinical_data_with_rna.loc[:, \"tumor_stage_pathological\"] == \"Stage III\", True, False)\n",
    "\n",
    "# create new clinical dataframe with mask\n",
    "stageI_clinical_with_rna = pd.DataFrame(clinical_data_with_rna[stageI_mask_with_rna])\n",
    "stageIII_clinical_with_rna = pd.DataFrame(clinical_data_with_rna[stageIII_mask_with_rna])\n",
    "\n",
    "# create new rna dataframe with mask\n",
    "stageI_rna = pd.DataFrame(rna_data[stageI_mask_with_rna])\n",
    "stageIII_rna = pd.DataFrame(rna_data[stageIII_mask_with_rna])\n",
    "\n",
    "# apply log2 to rna_data \n",
    "stageI_rna = np.log2(stageI_rna)\n",
    "stageIII_rna = np.log2(stageIII_rna)\n",
    "\n",
    "# drop NaN values and inf values\n",
    "stageI_rna = stageI_rna.replace(-np.inf, np.nan).replace(np.inf, np.nan).dropna(axis=1)\n",
    "stageIII_rna = stageIII_rna.replace(-np.inf, np.nan).replace(np.inf, np.nan).dropna(axis=1)\n",
    "\n",
    "# Create lists to store DE results\n",
    "top5_rna_list = [\"\", \"\", \"\", \"\", \"\"]\n",
    "rna_top5_DE = [0, 0, 0, 0, 0]\n",
    "\n",
    "# Find the top 5 DE RNA\n",
    "# loop through all RNA in rna_data\n",
    "for col in rna_data.columns:\n",
    "    # check if the RNA is in both comparison groups\n",
    "    if col in stageI_rna and col in stageIII_rna:\n",
    "        \n",
    "        # count expression\n",
    "        sum = 0\n",
    "        counter = 0\n",
    "        for row in stageI_rna.index:\n",
    "            sum += stageI_rna.loc[row, col]\n",
    "            counter += 1\n",
    "        stageI_mean_expression = sum / counter # calculate average\n",
    "        \n",
    "        # count expression\n",
    "        sum = 0\n",
    "        counter = 0\n",
    "        for row in stageIII_rna.index:\n",
    "            sum += stageIII_rna.loc[row, col]\n",
    "            counter += 1\n",
    "        stageIII_mean_expression = sum / counter #calculate average\n",
    "       \n",
    "        # find difference between average expression\n",
    "        differential_expression = abs(stageI_mean_expression - stageIII_mean_expression)\n",
    "       \n",
    "        # check if the RNA's differential expression is in the top 5 and adjust the lists if needed\n",
    "        if differential_expression > rna_top5_DE[4]:\n",
    "            if differential_expression > rna_top5_DE[3]:\n",
    "                if differential_expression > rna_top5_DE[2]:\n",
    "                    if differential_expression > rna_top5_DE[1]:\n",
    "                        if differential_expression > rna_top5_DE[0]:\n",
    "                            rna_top5_DE[4] = rna_top5_DE[3]\n",
    "                            top5_rna_list[4] = top5_rna_list[3]\n",
    "                            rna_top5_DE[3] = rna_top5_DE[2]\n",
    "                            top5_rna_list[3] = top5_rna_list[2]\n",
    "                            rna_top5_DE[2] = rna_top5_DE[1]\n",
    "                            top5_rna_list[2] = top5_rna_list[1]\n",
    "                            rna_top5_DE[1] = rna_top5_DE[0]\n",
    "                            top5_rna_list[1] = top5_rna_list[0]\n",
    "                            rna_top5_DE[0] = differential_expression\n",
    "                            top5_rna_list[0] = col\n",
    "                        else:\n",
    "                            rna_top5_DE[4] = rna_top5_DE[3]\n",
    "                            top5_rna_list[4] = top5_rna_list[3]\n",
    "                            rna_top5_DE[3] = rna_top5_DE[2]\n",
    "                            top5_rna_list[3] = top5_rna_list[2]\n",
    "                            rna_top5_DE[2] = rna_top5_DE[1]\n",
    "                            top5_rna_list[2] = top5_rna_list[1]\n",
    "                            rna_top5_DE[1] = differential_expression\n",
    "                            top5_rna_list[1] = col\n",
    "                    else:\n",
    "                        rna_top5_DE[4] = rna_top5_DE[3]\n",
    "                        top5_rna_list[4] = top5_rna_list[3]\n",
    "                        rna_top5_DE[3] = rna_top5_DE[2]\n",
    "                        top5_rna_list[3] = top5_rna_list[2]\n",
    "                        rna_top5_DE[2] = differential_expression\n",
    "                        top5_rna_list[2] = col\n",
    "                else:\n",
    "                    rna_top5_DE[4] = rna_top5_DE[3]\n",
    "                    top5_rna_list[4] = top5_rna_list[3]\n",
    "                    rna_top5_DE[3] = differential_expression\n",
    "                    top5_rna_list[3] = col\n",
    "            else:\n",
    "                rna_top5_DE[4] = differential_expression\n",
    "                top5_rna_list[4] = col\n",
    "\n",
    "# display results\n",
    "print('Top 5 DE RNAs: ', top5_rna_list)\n",
    "print('DE values: ', rna_top5_DE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "604f50ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top 5 DE proteins:  ['FTL', 'HBZ', 'HBA2', 'CMA1', 'HBB']\n",
      "DE values:  [0.8621652797318512, 0.6034531374509484, 0.5891002256474992, 0.5834638847268123, 0.5573026444527898]\n"
     ]
    }
   ],
   "source": [
    "# Find top 5 DE proteins\n",
    "\n",
    "# create masks for the two clinical feature groups\n",
    "stageI_mask = np.where(clinical_data.loc[:, \"tumor_stage_pathological\"] == \"Stage I\", True, False)\n",
    "stageIII_mask = np.where(clinical_data.loc[:, \"tumor_stage_pathological\"] == \"Stage III\", True, False)\n",
    "\n",
    "# create new clinical dataframe with mask\n",
    "stageI_clinical = pd.DataFrame(clinical_data[stageI_mask])\n",
    "stageIII_clinical = pd.DataFrame(clinical_data[stageIII_mask])\n",
    "\n",
    "# create new protein dataframe with mask\n",
    "stageI_protein = pd.DataFrame(protein_data[stageI_mask])\n",
    "stageIII_protein = pd.DataFrame(protein_data[stageIII_mask])\n",
    "\n",
    "# mask out infinity values\n",
    "stageI_negative_inf_mask = np.where(stageI_protein.loc[:,:] == float('-inf'), False, True)\n",
    "stageIII_negative_inf_mask = np.where(stageIII_protein.loc[:,:] == float('-inf'), False, True)\n",
    "stageI_protein = pd.DataFrame(stageI_protein[stageI_negative_inf_mask])\n",
    "stageIII_protein = pd.DataFrame(stageIII_protein[stageIII_negative_inf_mask])\n",
    "stageI_positive_inf_mask = np.where(stageI_protein.loc[:,:] == float('inf'), False, True)\n",
    "stageIII_positive_inf_mask = np.where(stageIII_protein.loc[:,:] == float('inf'), False, True)\n",
    "stageI_protein = pd.DataFrame(stageI_protein[stageI_positive_inf_mask])\n",
    "stageIII_protein = pd.DataFrame(stageIII_protein[stageIII_positive_inf_mask])\n",
    "\n",
    "# filter out NaN values\n",
    "stageI_protein = stageI_protein.dropna(axis=\"columns\")\n",
    "stageIII_protein = stageIII_protein.dropna(axis=\"columns\")\n",
    "\n",
    "# Create lists to store DE results\n",
    "top5_protein_list = [\"\", \"\", \"\", \"\", \"\"]\n",
    "protein_top5_DE = [0, 0, 0, 0, 0]\n",
    "\n",
    "# Find the top 5 DE protein\n",
    "# loop through all proteins in protein_data\n",
    "for col in protein_data.columns:\n",
    "    # check if the protein is in both comparison groups\n",
    "    if bool(col in stageI_protein) & bool(col in stageIII_protein):\n",
    "        \n",
    "        # count expression\n",
    "        sum = 0\n",
    "        counter = 0\n",
    "        for row in stageI_protein.index:\n",
    "            sum += stageI_protein.loc[row, col]\n",
    "            counter += 1\n",
    "        stageI_mean_expression = sum / counter # calculate average\n",
    "        \n",
    "        # count expression\n",
    "        sum = 0\n",
    "        counter = 0\n",
    "        for row in stageIII_protein.index:\n",
    "            sum += stageIII_protein.loc[row, col]\n",
    "            counter += 1\n",
    "        stageIII_mean_expression = sum / counter # calculate average\n",
    "\n",
    "        # find difference between average expression\n",
    "        differential_expression = abs(stageI_mean_expression - stageIII_mean_expression)\n",
    " \n",
    "        # check if the protein's differential expression is in the top 5 and adjust the lists if needed\n",
    "        try:\n",
    "            if differential_expression > protein_top5_DE[4]:\n",
    "                if differential_expression > protein_top5_DE[3]:\n",
    "                    if differential_expression > protein_top5_DE[2]:\n",
    "                        if differential_expression > protein_top5_DE[1]:\n",
    "                            if differential_expression > protein_top5_DE[0]:\n",
    "                                protein_top5_DE[4] = protein_top5_DE[3]\n",
    "                                top5_protein_list[4] = top5_protein_list[3]\n",
    "                                protein_top5_DE[3] = protein_top5_DE[2]\n",
    "                                top5_protein_list[3] = top5_protein_list[2]\n",
    "                                protein_top5_DE[2] = protein_top5_DE[1]\n",
    "                                top5_protein_list[2] = top5_protein_list[1]\n",
    "                                protein_top5_DE[1] = protein_top5_DE[0]\n",
    "                                top5_protein_list[1] = top5_protein_list[0]\n",
    "                                protein_top5_DE[0] = differential_expression\n",
    "                                top5_protein_list[0] = col\n",
    "                            else:\n",
    "                                protein_top5_DE[4] = protein_top5_DE[3]\n",
    "                                top5_protein_list[4] = top5_protein_list[3]\n",
    "                                protein_top5_DE[3] = protein_top5_DE[2]\n",
    "                                top5_protein_list[3] = top5_protein_list[2]\n",
    "                                protein_top5_DE[2] = protein_top5_DE[1]\n",
    "                                top5_protein_list[2] = top5_protein_list[1]\n",
    "                                protein_top5_DE[1] = differential_expression\n",
    "                                top5_protein_list[1] = col\n",
    "                        else:\n",
    "                            protein_top5_DE[4] = protein_top5_DE[3]\n",
    "                            top5_protein_list[4] = top5_protein_list[3]\n",
    "                            protein_top5_DE[3] = protein_top5_DE[2]\n",
    "                            top5_protein_list[3] = top5_protein_list[2]\n",
    "                            protein_top5_DE[2] = differential_expression\n",
    "                            top5_protein_list[2] = col\n",
    "                    else:\n",
    "                        protein_top5_DE[4] = protein_top5_DE[3]\n",
    "                        top5_protein_list[4] = top5_protein_list[3]\n",
    "                        protein_top5_DE[3] = differential_expression\n",
    "                        top5_protein_list[3] = col\n",
    "                else:\n",
    "                    protein_top5_DE[4] = differential_expression\n",
    "                    top5_protein_list[4] = col\n",
    "        except:\n",
    "            print('', end='')\n",
    "\n",
    "# display results\n",
    "print('Top 5 DE proteins: ', top5_protein_list)\n",
    "print('DE values: ', protein_top5_DE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "2a466db5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              AJAP1     DPEP1    GALNT5    IL20RB      RYR2       FTL  \\\n",
      "C3L-00004 -2.519207 -1.239811 -5.281123  2.685910 -0.721420 -1.724339   \n",
      "C3L-00010 -0.069282  4.515644 -1.685391  0.394540 -2.134346 -0.363228   \n",
      "C3L-00026  0.177869  4.635430  2.729468  1.139609 -1.565029  1.301930   \n",
      "C3L-00079 -1.161677  0.376214 -3.413559  0.441469 -0.421905 -1.496648   \n",
      "C3L-00088 -1.951992  0.384952 -0.749469  0.716994 -1.197209 -1.895355   \n",
      "...             ...       ...       ...       ...       ...       ...   \n",
      "C3N-01361  1.600187  1.924839  0.110036  0.307398 -1.436694 -0.461986   \n",
      "C3N-01522 -1.398691  2.166432 -6.167662  0.298801  2.159864 -1.675052   \n",
      "C3N-01646 -1.251963 -0.946572 -0.635367  7.399686 -0.645290 -1.817120   \n",
      "C3N-01649 -0.097150  3.354027 -1.442680 -0.220637  2.057367 -1.066512   \n",
      "C3N-01808 -0.590156  4.861775 -2.466955  0.540075 -1.611685  1.895297   \n",
      "\n",
      "                HBZ      HBA2      CMA1       HBB  \n",
      "C3L-00004 -0.370098 -0.786307 -0.918273 -0.762287  \n",
      "C3L-00010  0.240576 -0.081263  0.747860 -0.034944  \n",
      "C3L-00026  0.981930  0.780756  0.049086  0.855097  \n",
      "C3L-00079 -0.441854 -1.014088 -0.620829 -0.939448  \n",
      "C3L-00088  1.055353  0.750425 -1.555444  0.769956  \n",
      "...             ...       ...       ...       ...  \n",
      "C3N-01361  0.317262  0.428837  1.265030  0.481028  \n",
      "C3N-01522 -0.589292 -0.945236 -1.561464 -0.864699  \n",
      "C3N-01646 -0.724017 -1.180566 -0.000699 -1.127855  \n",
      "C3N-01649 -0.109630 -0.584516  0.322161 -0.527777  \n",
      "C3N-01808  2.056017  1.485439 -1.125322  1.538612  \n",
      "\n",
      "[85 rows x 10 columns]\n",
      "          tumor_stage_pathological\n",
      "C3L-00004                Stage III\n",
      "C3L-00010                  Stage I\n",
      "C3L-00026                  Stage I\n",
      "C3L-00079                Stage III\n",
      "C3L-00088                Stage III\n",
      "...                            ...\n",
      "C3N-01361                  Stage I\n",
      "C3N-01522                  Stage I\n",
      "C3N-01646                Stage III\n",
      "C3N-01649                Stage III\n",
      "C3N-01808                  Stage I\n",
      "\n",
      "[85 rows x 1 columns]\n"
     ]
    }
   ],
   "source": [
    "# Top 5 RNA: 'AJAP1', 'DPEP1', 'GALNT5', 'IL20RB', 'RYR2'\n",
    "# Top 5 Protein: 'FTL', 'HBZ', 'HBA2', 'CMA1', 'HBB'\n",
    "\n",
    "# apply log2 to rna_data \n",
    "rna_data_copy = pd.DataFrame(np.log2(rna_data))\n",
    "\n",
    "# drop NaN values and inf values\n",
    "rna_data_copy = rna_data_copy.replace(-np.inf, np.nan).replace(np.inf, np.nan).dropna(axis=1)\n",
    "\n",
    "rna_x = pd.DataFrame(rna_data_copy.loc[:,top5_rna_list])\n",
    "\n",
    "\n",
    "protein_data_copy = pd.DataFrame(protein_data)\n",
    "\n",
    "# drop NaN values and inf values\n",
    "protein_data_copy = protein_data_copy.replace(-np.inf, np.nan).replace(np.inf, np.nan).dropna(axis=1)\n",
    "\n",
    "protein_x = pd.DataFrame(protein_data_copy.loc[:,top5_protein_list])\n",
    "\n",
    "# create empty DataFrames for x_data and y_data for ML\n",
    "x_data = pd.DataFrame()\n",
    "y_data = pd.DataFrame()\n",
    "\n",
    "# fill x_data and y_data columns\n",
    "for rna in top5_rna_list:\n",
    "    x_data[rna] = []\n",
    "for protein in top5_protein_list:\n",
    "    x_data[protein] = []\n",
    "x_data['tumor_stage_pathological'] = []\n",
    "y_data['tumor_stage_pathological'] = []\n",
    "\n",
    "\n",
    "# add rows to fill in expression data for x_data and y_data\n",
    "\n",
    "# loop through all patient IDs\n",
    "for row in clinical_data.index:\n",
    "    # check if pateint ID in both protein_data and rna_data\n",
    "    if bool(row in protein_x.index) & bool(row in rna_x.index):\n",
    "        \n",
    "        # add row to fill in expression data for x_data\n",
    "        x_row_data = {\n",
    "                    top5_rna_list[0]:[rna_x.loc[row, top5_rna_list[0]]], \n",
    "                    top5_rna_list[1]:[rna_x.loc[row, top5_rna_list[1]]],\n",
    "                    top5_rna_list[2]:[rna_x.loc[row, top5_rna_list[2]]],\n",
    "                    top5_rna_list[3]:[rna_x.loc[row, top5_rna_list[3]]],\n",
    "                    top5_rna_list[4]:[rna_x.loc[row, top5_rna_list[4]]],\n",
    "                    top5_protein_list[0]:[protein_x.loc[row, top5_protein_list[0]]], \n",
    "                    top5_protein_list[1]:[protein_x.loc[row, top5_protein_list[1]]], \n",
    "                    top5_protein_list[2]:[protein_x.loc[row, top5_protein_list[2]]], \n",
    "                    top5_protein_list[3]:[protein_x.loc[row, top5_protein_list[3]]], \n",
    "                    top5_protein_list[4]:[protein_x.loc[row, top5_protein_list[4]]],\n",
    "                    'tumor_stage_pathological':[clinical_data.loc[row, 'tumor_stage_pathological']]\n",
    "                    }\n",
    "        \n",
    "        # add gene name to row\n",
    "        x_new_row = pd.DataFrame(x_row_data, index=[row])\n",
    "        x_data = pd.concat([x_data, x_new_row])\n",
    "        \n",
    "        # add row to fill in clinicial stage data for x_data\n",
    "        y_row_data = {\n",
    "                    'tumor_stage_pathological':[clinical_data.loc[row, 'tumor_stage_pathological']]\n",
    "                    }\n",
    "        # add gene name to rows\n",
    "        y_new_row = pd.DataFrame(y_row_data, index=[row])\n",
    "        y_data = pd.concat([y_data, y_new_row])\n",
    "\n",
    "\n",
    "# remove non-target data, inf/-inf, and NaN values\n",
    "x_data = x_data.replace('Stage II', np.nan).replace('Stage IV', np.nan).replace(-np.inf, np.nan).replace(np.inf, np.nan).dropna(axis=0)\n",
    "y_data = y_data.replace('Stage II', np.nan).replace('Stage IV', np.nan).replace(-np.inf, np.nan).replace(np.inf, np.nan).dropna(axis=0)\n",
    "\n",
    "# remove target from train data\n",
    "x_data = x_data.drop(columns='tumor_stage_pathological')\n",
    "\n",
    "print(x_data)\n",
    "print(y_data)       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "9a27967e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# scale the separated data\n",
    "scaler = StandardScaler()\n",
    "scaled_x_data = scaler.fit_transform(x_data)\n",
    "\n",
    "# convert y_data into a list\n",
    "y_data_list = []\n",
    "for row in y_data.index:\n",
    "    y_data_list.append(y_data.loc[row, 'tumor_stage_pathological'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "8acbb9d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Completed 2/10 simulations\n",
      "Completed 4/10 simulations\n",
      "Completed 6/10 simulations\n",
      "Completed 8/10 simulations\n",
      "Completed 10/10 simulations\n",
      "\n",
      "After 10 simulations, the average accuracy of each classifier is as follows:\n",
      "\tKNeighborsClassifier : 0.7269230769230769\n",
      "\tDecisionTreeClassifier : 0.6230769230769231\n",
      "\tMLPClassifier : 0.7269230769230769\n",
      "\tGaussianNB : 0.773076923076923\n"
     ]
    }
   ],
   "source": [
    "# Test four classifiers\n",
    "\n",
    "classifiers_names = [\n",
    "    'KNeighborsClassifier', \n",
    "    'DecisionTreeClassifier', \n",
    "    'MLPClassifier', \n",
    "    'GaussianNB'\n",
    "]\n",
    "\n",
    "classifiers = [\n",
    "    KNeighborsClassifier(),\n",
    "    DecisionTreeClassifier(),\n",
    "    MLPClassifier(),\n",
    "    GaussianNB()\n",
    "]\n",
    "\n",
    "classifiers_perf = {\n",
    "    0: [],\n",
    "    1: [],\n",
    "    2: [],\n",
    "    3: []\n",
    "}\n",
    "\n",
    "# run prediction models 10 times\n",
    "for n in range(10):\n",
    "    \n",
    "    # Train-test split\n",
    "    X_train, X_test, y_train, y_test = train_test_split(scaled_x_data, y_data_list, train_size=0.7)\n",
    "    \n",
    "    # run all four classifiers\n",
    "    for i in range(len(classifiers)):\n",
    "        model = classifiers[i]\n",
    "        model.fit(X_train, y_train)\n",
    "        y_pred = model.predict(X_test)\n",
    "        \n",
    "        # calculate accuracy of the model\n",
    "        correct = 0\n",
    "        index = 0\n",
    "        for length in y_pred:\n",
    "            if y_pred[index] == y_test[index]:\n",
    "                correct += 1\n",
    "            index += 1\n",
    "        accuracy = correct / len(y_test) \n",
    "        classifiers_perf[i].append(accuracy) # record accuracy\n",
    "\n",
    "    # display progress (intervals 0f two)\n",
    "    if ((n+1) % 2) == 0:\n",
    "        print(f'Completed {n+1}/10 simulations')\n",
    "\n",
    "# calculate average accuracies and print\n",
    "print('\\nAfter 10 simulations, the average accuracy of each classifier is as follows:')\n",
    "for i in classifiers_perf:\n",
    "    print(f'\\t{classifiers_names[i]} : {np.mean(classifiers_perf[i])}')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e7634d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# KNeighborsClassifier, MLPClassifier, and GaussianNB all seem to have an accuracy between 0.7 and 0.8 after 10 simuations.\n",
    "# DecisionTreeClassifier seems to be the lowest with an average accuracy between 0.55 and 0.7 "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
